tfimm/architectures/lora/README.md

```diff
@@ -0,0 +1,175 @@
+  class of each model is defined via the `cfg_class` class attribute.
+- The configuration of the LoRA model is a superset of the configuration of the base
+  model.
+
+Under these assumptions we can use the `register_lora_architecture` decorator to
+associate `LoRAResNet` as the LoRA variant of the `ResNet` class.
+
+Now, given an instance of `ResNet`, we can use `convert_to_lora_model` to convert
+it to a `LoRAResNet` instance _and_ transfer all weights.
+
+```python
+model = ResNet(cfg=ResNetConfig())
+... # Build model or load pre-trained weights
+
+lora_model = lora.convert_to_lora_model(model, lora_rank=2)
+```
+
+The `lora_model.trainable_weights` property correctly returns only the LoRA trainable
+weights, i.e., the low-rank updates. We additionally have the option to train the
+biases as well, either only for LoRA layers or for all layers. This can be specified
+by passing the values `"none"`, `"lora_only"` or  `"all"` for the `lora_train_bias` 
+parameter.
+
+```python
+lora_model = convert_to_lora_model(
+    model, lora_rank=2, lora_train_bias="lora_only"
+)
+```
+
+## Sequential and functional models
+
+The current implementations focusses on models created by subclassing, which is the case
+for all `tfimm` models. In particular, the registry system works only for subclassed
+models. However, some of the functionality also works for functional models.
+
+- The LoRA layers are the basic building blocks for both subclassed as well as 
+  functional models.
+- Transferring weights works for all models, regardless of type, provided the regular
+  model and LoRA variant have the same architecture with the exception of LoRA layers.
+  Use the function to tranfer weights to LoRA.
+  ```python
+  from tfimm.architectures import lora
+  from tfimm.models import transfer_weights
+  
+  # Transfer weights into the LoRA model
+  transfer_weights(
+      regular_model, lora_model, weights_to_ignore=lora.LORA_WEIGHT_NAMES)
+  ```
+- After training, we need to manually merge weights and then transfer them back to the
+  regular model.
+  ```python
+  lora.merge_weights(lora_model)
+  transfer_weights(lora_model, regular_model)
+  ```
+- The functions `lora_trainable_weights` and `lora_non_trainable_weights` work for all
+  models, regardless of type and return a list of weights to be used for LoRA training
+  (or all other weights).
+
+## Comparison to PyTorch implementation
+
+The official PyTorch implementation can be found 
+[here](https://github.com/microsoft/LoRA/tree/main). The differences between TensorFlow
+and PyTorch mean that we cannot simply replicate the PyTorch implementation.
+
+- In PyTorch it is possible to mark only LoRA parameters as trainable via
+  ```python
+  # This sets requires_grad to False for all parameters without the 
+  # string "lora_" in their names
+  lora.mark_only_lora_as_trainable(model)
+  ```
+  In TensorFlow it is _not_ possible to change the `trainable` attribute of a variable
```

Comment:
Is it worth mentioning this issue here: https://github.com/tensorflow/tensorflow/issues/47597 ?

Reply:
Yes, thx.